### PR TITLE
Add dark theme to Better Informatics

### PR DESCRIPTION
**Pull request by yellowtides**
## Light theme
Screenshot 2021-06-10 at 23 13 20

## Dark theme
Screenshot 2021-06-10 at 23 13 12

## Functionality

There is now a little switch (handdrawn by yours truly) under the navigation bar of every `frontpage.html`. It can be can pressed to toggle between themes.

## Changelog

1. New javascript file under `/static/js` that is loaded from `head.html`, just like with the existing static stylesheet. Contains a simple check of the user's preferred theme. Also defines the switch's functionality.

2. Defined extra styling for dark mode and the switch element. The palette is heavily inspired by https://sigint.mx/ 's.

3. The biggest change is probably in the image assets (`/static/img`). Instead of the previous `color.png` and `gray.png`, there is now `color-light.png`, `color-dark.png`, `gray-light.png`, and `gray-dark.png`. Instead of `betterinf-title`, there is now `betterinf-title-dark.png` and `betterinf-title-light.png`. The images now have a transparent background and the colors used are adjusted based on the referenced theme (e.g., dark mode has a lighter font color).

## Bumps

1. I opted for a standard js solution that stores the user's preferred theme as a `localStorage` variable. Unfortunately, this means that dark mode users experience a flash of light when moving inbetween pages, as the javascript executes after the default (lightmode) stylesheet is loaded. Light mode users are spared.
2. Pressing the switch to go from darkmode to lightmode makes the currently selected monkey/human (which represents the current page) in the navigation bar to slightly lag behind the others. This is due to the `0.3s transition` that styles every `#menu a` element.

### CODE REVIEW

**Comment by yutotakano** [comment]
May I propose you switch to the `prefers-color-scheme` media query? 

image

Points of consideration:
- No need for client to manually toggle, as it detects the browser (and the desktop)'s preference automatically
- It removes the need for javascript, resulting a slightly faster pageload, plus reducing any white-to-dark flickering upon page load
- Browser support is good (https://caniuse.com/?search=prefers-color-scheme), and if it doesn't support the query it simply defaults to the default (white) version.

Testing can be done with Developer Tools' Rendering tab.
image

**Comment by yellowtides** [comment]
I intially considered the above solution, but decided against it due to two main points of consideration: the lack of backwards compatibility with some of the browsers on the list you posted, and my suspicion that there exist a non-zero number of dark mode users that do not run dark mode system-wide.

**Comment by 9nine9nine9** [comment]
As pointed out, by default you would set a light theme and override with a dark theme if requested, so it would be compatible with old browsers regardless.

There's a nice implementation on [gwern's site](https://www.gwern.net/Design) (if you search for dark mode, you can see their arguments on this matter btw), and the js to do this is [public domain](https://www.gwern.net/static/js/darkmode.js). I believe this solution fits all use cases described above?

**Comment by yellowtides** [comment]
Fair enough, made all the changes necessary. Looks cleaner! :D

**Review by qaisjp — COMMENTED** [review]
- I wonder if we can use this opportunity to convert the title to text instead of an image (with a font? either figure out the font or use comic sans lul)
- can you resize the switch? it looks much larger than necessary
- can you compress all images using zopfli? or something else if zopfli is too complicated

**Comment by qaisjp** [comment]
Also the switch is quite wasteful of vertical space on that page (but maybe isn't a big deal on other pages since there are buttons). I wonder if it would look good in the corner of the screen? `top: 0; right: blahblah`

Also, I think your PR is missing some code that uses the switch png?

**Comment by 9nine9nine9** [comment]
On the topic of image compressing, I was incidentally looking a few minutes ago at some github integrations that can optimise any images committed to the repo; might be useful to have?

[Here](https://github.com/marketplace/imgbot)'s one, though I'm sure there's plenty of alternatives.

**Comment by yellowtides** [comment]
Whoops, forgot to take the switches out @qaisjp. The reason why they're not used is because I've transitioned to the `@media` query solution proposed by @yutotakano in the meantime.

Re: image compression, I'm afraid of doing it as I think they look quite low-res as is (?)

**Comment by yellowtides** [comment]
^ Some more clean-up, knew I was forgetting something 😆

**Comment by yellowtides** [comment]
## New (switch-less, @media-ful look)

1. Dark mode
Screenshot 2021-06-11 at 03 31 40
2. Light mode
Screenshot 2021-06-11 at 03 31 57
3. Mobile dark
Screenshot 2021-06-11 at 03 33 53

**Comment by yellowtides** [comment]
All done! Only the comic sans bit remains...

**Comment by qaisjp** [comment]
Cool! I think you can refer to https://bettermathematics.github.io/ as they ended up converting the title to html

**Comment by yellowtides** [comment]
Unless I'm missing something, their title seems to be an image too:
BM logo

However, the font they use is similar but not quite the same. For reference,
BI logo

I can't find the exact font we use for BI, do you have any leads? If not, can I just use a strong slim serif lookalike like they have?

**Comment by pkage** [comment]
My best guesses for the fonts:

Better Mathematics: [Arctic 2 or 2.5](https://www.fontzillion.com/fonts/nathan-caldecott/arctik?utm_source=fontsquirrel.com&utm_medium=matcherator_link&utm_campaign=arctik) based on the low crossbar on the A but not the H

Better Informatics: not sure, I couldnt find a good match. Looks like @qaisjp [checked it in](https://github.com/compsoc-edinburgh/betterinformatics/commit/bf66bbac4f8685af34c82ac47028cff76d318684) many moons ago, maybe he knows more...?

Ultimately, it's kind of a moot point—I'm not a super stickler for getting every detail perfect, and if you want to replace the logo with a similar font that's A-OK by me. Just write down what font you used or commit the .psd lol

**Comment by yutotakano** [comment]
Sorry, I replied on Discord but forgot to comment here: I was able to identify the font as [Jolly](https://fonts2u.com/jolly.font)

**Comment by yellowtides** [comment]
Jolly seems to be free for personal use so I just included it as a .ttf file under /static/fonts. Removed the banner images, happy to say it looks the same! Thanks @yutotakano for being a legendary investigator as always and finding the exact font.

File: static/css/stylesheet.css [hunk 1]
```diff
@@ -379,11 +363,23 @@ h1, h2, h3, h4, h5, h6 {
...
     background-position: -822px 0
 }
+
+@font-face {
+    font-family: jolly;
+    src: url("../fonts/Jolly.ttf");
+}
+
 #title {
+    font-family: jolly;
```
Comment (qaisjp) [inline static/css/stylesheet.css:373]: maybe fall back on comic sans while the font loads? not sure if comic sans ms is actually available on non-windows platforms. not blocking though, you can commit this later

**Review by qaisjp — APPROVED** [review]
there are a lot of images in this branch so please squash merge :)

**Comment by qaisjp** [comment]
> Looks like @qaisjp checked it in many moons ago, maybe he knows more...?

I just copied it off the original Better Informatics website :P

**Pull request merged**